### PR TITLE
Improve table.rs

### PR DESCRIPTION
**Pull request by ddragana**
- define more descriptive errors,
- remove QPackSide,
-  do not use acked_inserts_cnt on the decoder side,
-  add comments,
-  define some struct pub(crate),
-  and some other small improvements.

### CODE REVIEW

**Comment by ddragana** [comment]
This is based on #576

**Comment by ddragana** [comment]
This is part 1 fro fixing #539.

File: neqo-qpack/src/encoder.rs [hunk 1]
```diff
@@ -438,6 +460,36 @@ mod tests {
         }
     }
 
+    fn connect_flow_control(max_data: u64) -> TestEncoder {
```
Comment (martinthomson) [inline neqo-qpack/src/encoder.rs:463]: This is a lot of duplicate code.  Maybe you should have a baseline function like:

```rust
fn connect_generic<F>(f: F) -> TestEncoder 
where F : FnOnce(&mut Connection,&mut Connection) {
        let mut conn = default_client();
        let mut peer_conn = default_server(); 
        f(&mut conn, &mut peer_conn);
        
         // ...
}
```
This could then be:
```rust
fn connect_fc(max_data: u64) -> TestEncoder {
  connect_generic(|client, server| {
     server.set_local_tparam(...);
  })
```

File: neqo-transport/src/connection.rs [hunk 2]
```diff
@@ -1478,6 +1478,11 @@ impl Connection {
         }
     }
 
+    // TODO: this should be #[cfg(test)], but it is not acceptable from a different crate.
+    pub fn handle_max_data_test(&mut self, maximum_data: u64) {
+        self.handle_max_data(maximum_data);
+    }
```
Comment (martinthomson) [inline neqo-transport/src/connection.rs:1484]: We need a different way to handle this.

What should be happening in your test is that the client hits the flow control limit and sends a STREAM_DATA_BLOCKED frame.  The server should respond by increasing the flow control limits.  That should suffice for your test, meaning that you don't need this function at all.

File: neqo-qpack/src/table.rs [hunk 3]
```diff
@@ -57,63 +58,67 @@ impl DynamicTableEntry {
 }
 
 #[derive(Debug)]
-pub struct HeaderTable {
-    qpack_side: QPackSide,
+pub(crate) struct HeaderTable {
```
Comment (martinthomson) [inline neqo-qpack/src/table.rs:61]: Is this change necessary?  That is, is there anything in this module that needs to be used outside of the crate?

File: neqo-http3/src/transaction_client.rs [hunk 4]
```diff
@@ -54,19 +54,25 @@ impl Request {
         r
     }
 
-    fn ensure_encoded(&mut self, encoder: &mut QPackEncoder, stream_id: u64) {
+    fn ensure_encoded(
```
[the remark this answers is not in the document]
Comment (ddragana) [reply neqo-http3/src/transaction_client.rs:57]: fine by me, I will change it.

File: neqo-transport/src/connection.rs (hunk 2 above)
Comment (ddragana) [reply neqo-transport/src/connection.rs:1484]: I will leave it in this PR and change it #599.

File: neqo-qpack/src/table.rs (hunk 3 above)
Comment (ddragana) [reply neqo-qpack/src/table.rs:61]: No, there is nothing that uses this outside crate, that is why I limited its scope.

**Comment by ddragana** [comment]
> There are lots of `map_err` calls in this change. It seems that the header table itself can't be the final decider on what the error code should be. Is it possible to build a single error code mapping function at a higher layer? It might need to be slightly smarter (i.e., use `match` to limit what errors it changes), but it might help contain this proliferation of error code mappings that seem brittle and hard to maintain.

Previously (long time ago) I tried to make decision in table.rs, but that was not working very nice.
encoder, decoder and header_block must set the final error.
I find it nicer to explicitly map errors, because the mapping is usually per function, e.g. execute_instruction will always return only EncoderStream.
I think I can reduce number of map_err or even have a mapping function, but I need to look at decoder.rs and encoder.rs first. This is just one step in making neqo-qpack final. I did not want to make a huge PR therefore this split, small steps. We can leave this open am merge it with other PR that are coming.

File: neqo-transport/src/connection.rs (hunk 2 above)
Comment (ddragana) [reply neqo-transport/src/connection.rs:1484]: #599 has been updated.
